datafusion-physical-expr/src/expressions/approx_percentile_cont.rs

```diff
@@ -194,75 +204,125 @@ pub struct ApproxPercentileAccumulator {
 impl ApproxPercentileAccumulator {
     pub fn new(percentile: f64, return_type: DataType) -> Self {
         Self {
-            digest: TDigest::new(100),
+            digest: TDigest::new(DEFAULT_MAX_SIZE),
             percentile,
             return_type,
         }
     }
-}
 
-impl Accumulator for ApproxPercentileAccumulator {
-    fn state(&self) -> Result<Vec<ScalarValue>> {
-        Ok(self.digest.to_scalar_state())
+    pub(crate) fn merge_digests(&mut self, digests: &[TDigest]) {
+        self.digest = TDigest::merge_digests(digests);
     }
 
-    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
-        debug_assert_eq!(
-            values.len(),
-            1,
-            "invalid number of values in batch percentile update"
-        );
-        let values = &values[0];
-
-        self.digest = match values.data_type() {
+    pub(crate) fn convert_to_ordered_float(
+        values: &ArrayRef,
+    ) -> Result<Vec<OrderedFloat<f64>>> {
+        match values.data_type() {
             DataType::Float64 => {
                 let array = values.as_any().downcast_ref::<Float64Array>().unwrap();
-                self.digest.merge_unsorted(array.values().iter().cloned())?
+                Ok(array
+                    .values()
+                    .iter()
+                    .filter_map(|v| v.try_as_f64().transpose())
+                    .collect::<Result<Vec<_>>>()?)
```

Comment:
originally, I want to just return `array.values().iter().cloned()` and then pass to `merge_sorted` as `impl IntoIterator<Item = T>`, but I didn't figure out how to correct `type` it, so I change to `Vec<OrderedFloat<f64>>`


https://github.com/apache/arrow-datafusion/blob/485bfa6ffcf8dc836a482127ecd367d9abfa7248/datafusion-physical-expr/src/tdigest/mod.rs#L252-L255

Reply:
Yeah, it is tough because the type of the various branches are different.

You could could save the copy by doing something like letting the caller provide a function that gets invoked for each element (untested):

```rust
    pub(crate) fn convert_to_ordered_float(
        values: &ArrayRef,
        f: impl FnMut(Option<OrderedFloat<f64>))
    ) ->  Result<()> {
```

And then call `f()` on each element;

```rust
...
            DataType::Float32 => {
                let array = values.as_any().downcast_ref::<Float32Array>().unwrap();
                array
                    .values()
                    .iter()
                    .try_for_each(|v| {
                      f(v.try_as_f64()?)
                     })
            }
...
```